internal/plugincommon/discovery/processor.go

```diff
@@ -0,0 +1,120 @@
+package discovery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugincommon"
+	dt "github.com/smartcontractkit/chainlink-ccip/internal/plugincommon/discovery/discoverytypes"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/consts"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/reader"
+)
+
+// Ensure that PluginProcessor is implemented.
+var _ plugincommon.PluginProcessor[dt.Query, dt.Observation, dt.Outcome] = &ContractDiscoveryProcessor{}
+
+// ContractDiscoveryProcessor is a plugin processor for discovering contracts.
+type ContractDiscoveryProcessor struct {
+	lggr      logger.Logger
+	reader    *reader.CCIPReader
+	homechain reader.HomeChain
+	dest      cciptypes.ChainSelector
+	bigF      int
+}
+
+func NewContractDiscoveryProcessor(
+	lggr logger.Logger,
+	reader *reader.CCIPReader,
+	homechain reader.HomeChain,
+	dest cciptypes.ChainSelector,
+	bigF int,
+) *ContractDiscoveryProcessor {
+	return &ContractDiscoveryProcessor{
+		lggr:      lggr,
+		reader:    reader,
+		homechain: homechain,
+		dest:      dest,
+		bigF:      bigF,
+	}
+}
+
+// Query is not needed for this processor.
+func (cdp *ContractDiscoveryProcessor) Query(_ context.Context, _ dt.Outcome) (dt.Query, error) {
+	return nil, nil
+}
+
+// Observation reads contract addresses from the destination chain.
+// In the future this should be extended to omit observations unless one of the nodes requests addresses.
+func (cdp *ContractDiscoveryProcessor) Observation(
+	ctx context.Context, _ dt.Outcome, _ dt.Query,
+) (dt.Observation, error) {
+	contracts, err := (*cdp.reader).DiscoverContracts(ctx, cdp.dest)
```

Comment:
If you remove the pointer-to-interface this can be simplified to


```suggestion
	contracts, err := cdp.reader.DiscoverContracts(ctx, cdp.dest)
```